map_creator/src/find_base_byGraspReachabilityScore.cpp

```diff
@@ -0,0 +1,299 @@
+    grasp_pose2.orientation.w = grasp_quat2[3];
+    grasp_poses.push_back(grasp_pose2);   
+
+    geometry_msgs::Pose grasp_pose3;
+    grasp_pose3.position.x = 0.99;
+    grasp_pose3.position.y = 0.23;
+    grasp_pose3.position.z = 0.34;
+    grasp_pose3.orientation.x = grasp_quat3[0];
+    grasp_pose3.orientation.y = grasp_quat3[1];
+    grasp_pose3.orientation.z = grasp_quat3[2];
+    grasp_pose3.orientation.w = grasp_quat3[3];
+    grasp_poses.push_back(grasp_pose3);
+
+    multiset<pair<double, vector<double> > > scoreWithSp;
+    for (map<vector<double>, double > ::iterator it = SphereCol.begin();it != SphereCol.end();++it)
+    {
+      scoreWithSp.insert(pair<double, vector<double> >(it->second, it->first ));
+    }
+
+    ROS_INFO("Numer of Spheres : %lu",scoreWithSp.size());
+    
+    vector<vector<double> > highScoreSp;
+    for (multiset<pair<double, vector<double> > > ::reverse_iterator it = scoreWithSp.rbegin();it != scoreWithSp.rend();++it)
+    {
+	 highScoreSp.push_back(it->second);
+
+    }
+
+    ROS_INFO("Size of high score sp: %lu", highScoreSp.size()); //filtered sp centers based on their scores
+    vector<geometry_msgs::Pose> probBasePoses;
+
+   int numofSp = 10;//From how many spheres we are collecting all the poses
+    for(int i=0;i<numofSp;++i) 
+    {
+      multimap<vector<double>, vector<double> >::iterator it;
+      for(it = PoseColFilter.lower_bound(highScoreSp[i]); it !=PoseColFilter.upper_bound(highScoreSp[i]); ++it){
+	geometry_msgs::Pose pp;
+        sd.convertVectorToPose(it->second,pp); 
+        probBasePoses.push_back(pp);
+
+      }
+    }
+
+    cout<<"Size of Probable Base poses: "<<probBasePoses.size()<<" with "<<numofSp<<"Sphere"<<endl;
+
+    multiset<pair<int, vector<double> > >basePoseWithHits;
+    
+    for(int i=0;i<probBasePoses.size();++i)
+    {
+      int numofHits = 0;
+      for(int j=0;j<grasp_poses.size();++j)
+      {
+        int nsolns = 0;
+	numofHits+=isIkSuccesswithTransformedBase(probBasePoses[i],grasp_poses[j], nsolns);
+        
+      }
+	//cout<<numofHits<<endl;
+	vector<double> baseP;
+        sd.convertPoseToVector(probBasePoses[i], baseP);
+	basePoseWithHits.insert(pair<int, vector<double> >(numofHits , baseP));
+        //scoreWithSp.insert(pair<double, vector<double> >(it->second, it->first ));
+    }
+    
+
+
+    vector<geometry_msgs::Pose> final_base_poses;
+    for (multiset<pair<int, vector<double> > > ::iterator it = basePoseWithHits.begin();it != basePoseWithHits.end();++it)
+    {
+	//cout<<"Pose: "<<it->second[0]<<it->second[1]<<it->second[2]<<"=> has "<<it->first<<" solutions." <<endl;
+	if((it->first)>=grasp_poses.size())
+	{
+	  geometry_msgs::Pose final_base;
+	  sd.convertVectorToPose(it->second, final_base);
+          final_base_poses.push_back(final_base);
+	}
+
+    }
+
+    cout<<"# of final base poses: "<<final_base_poses.size()<<endl;
+
+ int numofFinalPose = 10;//How many optimal base pose we need 
```

Comment:
We are taking all the poses from the highest scoring spheres (The number of high scoring sphere->numofSp) which have the ability to reach all the grasp poses (graspscoring)/ providing max number of Ik solutions for the grasp poses (IKscoring) in a different container. So the pose collecion can be from the first sphere 20 (reachable) our of 100(total number of poses in that sphere) ....from a second sphere 10(reachable) out of 50(total number of poses)..In the new container now we have total 20+10= 30 poses that can reach all the grasp poses. We take 10(numofFinalPoses) from them.